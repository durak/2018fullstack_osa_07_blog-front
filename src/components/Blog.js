import PropTypes from 'prop-types'
import React from 'react'

class Blog extends React.Component {
  static propTypes = {
    handleDestroy: PropTypes.func.isRequired,
    handleLike: PropTypes.func.isRequired,
    blog: PropTypes.object.isRequired,
    user: PropTypes.object.isRequired
  }

  constructor(props) {
    super(props)
    this.state = {
      fullView: false
    }

  }

  toggleVisibility = () => {
    this.setState({ fullView: !this.state.fullView })
  }

  like = () => {
    let blog = { ...this.props.blog }

    blog.likes = blog.likes + 1
    this.props.handleLike(blog)
  }

  destroy = () => {
    let blog = this.props.blog
    if (window.confirm(`delete ${blog.title} by ${blog.author}?`)) {
      this.props.handleDestroy(blog)
    }
  }

  render() {
    const blog = this.props.blog

    const blogStyle = {
      paddingTop: 10,
      paddingLeft: 2,
      border: 'solid',
      borderWidth: 1,
      marginBottom: 5
    }

    const hideWhenVisible = { display: this.state.fullView ? 'none' : '' }
    const showWhenVisible = { display: this.state.fullView ? '' : 'none' }

    const destroyVisible =
      { display: blog.user && (blog.user.username !== this.props.user.username) ? 'none' : '' }


    return (
      <div style={blogStyle} >
        <div style={hideWhenVisible} className="minimized">
          <p className="clickable" onClick={this.toggleVisibility}>{blog.title} {blog.author}</p>
        </div>
        <div style={showWhenVisible} className="maximized">
          <p className="clickable" onClick={this.toggleVisibility}>{blog.title} {blog.author}</p>
          <a href={blog.url}>{blog.url}</a>
          <p>
            likes: {blog.likes}
            <button onClick={this.like}>like</button>
          </p>
          <p className="addedBy">added by {blog.user ? blog.user.name : 'anonymous'}</p>
          <div className="destroy" style={destroyVisible}>
            <p><button onClick={this.destroy}>delete</button></p>
          </div>
        </div>
      </div>
    )
  }
}

/* const Blog = ({ blog }) => (
  <div>
    {blog.title} {blog.author}
  </div>
) */

export default Blog