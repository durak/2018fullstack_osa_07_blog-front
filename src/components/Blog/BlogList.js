import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { Card } from 'semantic-ui-react'

import { blogLike } from '../../reducers/blogReducer'
import { notify } from '../../reducers/notificationReducer'

import BlogListItem from './blogList/BlogListItem'

export const BlogList = ({ blogs, blogLike, notify }) => {

  return (
    <div>
      <h1>Blogs</h1>
      <Card.Group centered itemsPerRow="4" stackable>
        {blogs.map(blog =>
          <BlogListItem
            key={blog.id}
            blog={blog}
            blogLike={blogLike}
            notify={notify}
          />
        )}
      </Card.Group>
    </div>
  )

}

const mapStateToProps = (state) => {
  let sortedBlogs = []
    .concat(state.blogs)
    .sort((a, b) => {
      return b.likes - a.likes
    })

  return {
    blogs: sortedBlogs
  }
}

export default connect(
  mapStateToProps,
  { blogLike, notify }
)(BlogList)

BlogListItem.propTypes = {
  blogs: PropTypes.arrayOf(PropTypes.object),
  blogLike: PropTypes.func.isRequired,
  notify: PropTypes.func.isRequired
}