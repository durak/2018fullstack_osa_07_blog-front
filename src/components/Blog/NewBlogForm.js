import PropTypes from 'prop-types'
import React from 'react'
import { connect } from 'react-redux'
import { Redirect } from 'react-router-dom'
import { Form, Button, Input } from 'semantic-ui-react'

import { blogCreate } from '../../reducers/blogReducer'
import { notify } from '../../reducers/notificationReducer'

class NewBlogForm extends React.Component {
  static propTypes = {
    blogCreate: PropTypes.func.isRequired,
    notify: PropTypes.func.isRequired
  }

  constructor(props) {
    super(props)
    this.state = {
      visible: false,
      title: '',
      author: '',
      url: '',
      valid: {
        title: true,
        author: true,
        url: true
      },
      redirectAfterSubmit: null
    }
  }

  componentDidUpdate(prevProps, prevState) {
    if (prevState.redirectAfterSubmit !== this.state.redirectAfterSubmit) {
      this.setState({ redirectAfterSubmit: null })
    }
  }

  validateInputs = () => {
    const valid = {
      title: this.state.title.length > 0,
      author: this.state.author.length > 0,
      url: this.state.url.length > 0
    }

    this.setState({
      valid
    })

    return valid.title && valid.author && valid.url
  }

  handleChange =  (event) => {
    this.setState({
      [event.target.name]: event.target.value
    })
  }

  handleSubmit = async (e) => {
    e.preventDefault()
    if (!await this.validateInputs()) return

    const newBlog = {
      title: this.state.title,
      author: this.state.author,
      url: this.state.url
    }

    const response = await this.props.blogCreate(newBlog)
    this.props.notify(`a new blog ${newBlog.title} by ${newBlog.author} added`, 'message')
    this.props.togglable.current.toggleVisibility()
    console.log('RESPOEEEEENNENSNSNNENE', response)

    this.setState({
      title: '', author: '', url: '',
      valid: { title: true, author: true, url: true },
      redirectAfterSubmit: response.id
    })
  }


  render() {

    if (this.state.redirectAfterSubmit) {
      return(<Redirect push to={`/blogs/${this.state.redirectAfterSubmit}`} />)
    }


    return (

      <Form onSubmit={this.handleSubmit}>
        <Form.Group widths="equal">
          <Form.Field>
            <label>title</label>
            <Form.Input
              fluid
              type="text"
              name="title"
              value={this.state.title}
              onChange={this.handleChange}
              error={!this.state.valid.title}
            />
          </Form.Field>

          <Form.Field>
            <label>author</label>
            <Form.Input
              fluid
              type="text"
              name="author"
              value={this.state.author}
              onChange={this.handleChange}
              error={!this.state.valid.author}
            />
          </Form.Field>

          <Form.Field>
            <label>url</label>
            <Form.Input
              fluid
              type="text"
              name="url"
              value={this.state.url}
              onChange={this.handleChange}
              error={!this.state.valid.url}
            />
          </Form.Field>

          <Button type="submit">submit</Button>
        </Form.Group>
      </Form>

    )
  }
}

export default connect(
  null,
  { blogCreate, notify }
)(NewBlogForm)