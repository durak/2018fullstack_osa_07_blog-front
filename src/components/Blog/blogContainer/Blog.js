import React from 'react'
import { Redirect } from 'react-router-dom'
import { Button } from 'semantic-ui-react'

const Blog = ({ blog, user, handleLike, handleDestroy }) => {

  if (blog === undefined) return <Redirect to="/" />

  const blogStyle = {
    paddingTop: 10,
    paddingLeft: 2,
    border: 'solid',
    borderWidth: 1,
    marginBottom: 5
  }

  const destroyButtonVisible = { display: blog.user && (blog.user.username !== user.username) ? 'none' : '' }

  return (
    <div style={blogStyle} >

      <h1>{blog.title}</h1>
      <h2>by: {blog.author}</h2>
      <a href={blog.url}>{blog.url}</a>

      <p>
        likes: {blog.likes}
        <Button onClick={handleLike}>like</Button>
      </p>

      <div className="destroy" style={destroyButtonVisible}>
        <p><Button onClick={handleDestroy}>delete</Button></p>
      </div>

      <p className="addedBy">added by {blog.user ? blog.user.name : 'anonymous'}</p>

    </div>
  )
}


export default Blog