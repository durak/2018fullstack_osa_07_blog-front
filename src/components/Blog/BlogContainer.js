import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'

import { blogLike, blogDestroy, blogAddComment } from '../../reducers/blogReducer'
import { notify } from '../../reducers/notificationReducer'

import Blog from './blogContainer/Blog'
import BlogComments from './blogContainer/BlogComments'
import BlogCommentForm from './blogContainer/BlogCommentForm'

export const BlogContainer = ({
  user,
  blogId,
  blogs,
  blogLike,
  blogDestroy,
  blogAddComment,
  notify }) => {

  const blog = blogs.find((b) => b.id === blogId)

  const handleDestroy = () => {
    if (window.confirm(`delete ${blog.title} by ${blog.author}?`)) {
      blogDestroy(blog)
      notify(`Blog ${blog.title} by ${blog.author} deleted`, 'message')
    }
  }

  const handleLike = () => {
    const msg = `you liked ${blog.title} by ${blog.author}`
    blogLike(blog)
    notify(msg, 'message')
  }

  const handleComment = (comment) => {
    const msg = `comment '${comment}' added to blog '${blog.title}'`
    blogAddComment(blog, comment)
    notify(msg, 'message')
  }


  return (
    <div>
      <Blog
        blog={blog}
        user={user}
        handleLike={handleLike}
        handleDestroy={handleDestroy}
      />
      <BlogComments blog={blog} />
      <BlogCommentForm handleComment={handleComment} />
    </div>
  )
}


const mapStateToProps = (state) => {
  return {
    blogs: state.blogs,
    user: state.loggedIn
  }
}

export default connect(
  mapStateToProps,
  { blogLike, blogDestroy, blogAddComment, notify }
)(BlogContainer)

BlogContainer.propTypes = {
  user: PropTypes.object.isRequired,
  blogId: PropTypes.string.isRequired,
  blogs: PropTypes.arrayOf(PropTypes.object).isRequired,
  blogLike: PropTypes.func.isRequired,
  blogAddComment: PropTypes.func.isRequired,
  notify: PropTypes.func.isRequired
}