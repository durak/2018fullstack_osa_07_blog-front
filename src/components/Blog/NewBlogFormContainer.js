import PropTypes from 'prop-types'
import React from 'react'
import { connect } from 'react-redux'
import { Redirect } from 'react-router-dom'
import { Sidebar, Segment } from 'semantic-ui-react'

import NewBlogForm from './newBlogFormContainer/NewBlogForm'
import { blogCreate } from '../../reducers/blogReducer'
import { notify } from '../../reducers/notificationReducer'
import { toggleSidebar } from '../../reducers/sidebarReducer'

class NewBlogFormContainer extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      title: '',
      author: '',
      url: '',
      valid: {
        title: true,
        author: true,
        url: true
      },
      redirectAfterSubmit: null
    }
  }

  componentDidUpdate(prevProps, prevState) {
    if (prevState.redirectAfterSubmit !== this.state.redirectAfterSubmit) {
      this.setState({ redirectAfterSubmit: null })
    }
  }

  validateInputs = () => {
    const valid = {
      title: this.state.title.length > 0,
      author: this.state.author.length > 0,
      url: this.state.url.length > 0
    }

    this.setState({
      valid
    })

    return valid.title && valid.author && valid.url
  }

  handleChange =  (event) => {
    this.setState({
      [event.target.name]: event.target.value
    })
  }

  handleSubmit = async (e) => {
    e.preventDefault()
    if (!await this.validateInputs()) return

    const newBlog = {
      title: this.state.title,
      author: this.state.author,
      url: this.state.url
    }

    const response = await this.props.blogCreate(newBlog)
    this.props.notify(`a new blog ${newBlog.title} by ${newBlog.author} added`, 'message')


    this.setState({
      title: '', author: '', url: '',
      valid: { title: true, author: true, url: true },
      redirectAfterSubmit: response.id
    })

    this.props.toggleSidebar()
  }

  render() {

    if (this.state.redirectAfterSubmit) {
      return(<Redirect push to={`/blogs/${this.state.redirectAfterSubmit}`} />)
    }


    return (
      <Sidebar as={Segment} animation="overlay" direction="bottom" visible={this.props.sidebar.visible}>
        <NewBlogForm
          handleSubmit={this.handleSubmit} handleChange={this.handleChange} handleCancel={this.props.toggleSidebar}
          title={this.state.title} author={this.state.author} url={this.state.url} valid={this.state.valid}
        />
      </Sidebar>
    )
  }
}

const mapStateToProps = (state) => {
  return {
    sidebar: state.sidebar
  }
}

export default connect(
  mapStateToProps,
  { blogCreate, notify, toggleSidebar }
)(NewBlogFormContainer)

NewBlogFormContainer.propTypes = {
  blogCreate: PropTypes.func.isRequired,
  notify: PropTypes.func.isRequired,
  toggleSidebar: PropTypes.func.isRequired,
  sidebar: PropTypes.object.isRequired
}