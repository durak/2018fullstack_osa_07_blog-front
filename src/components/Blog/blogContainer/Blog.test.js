import React from 'react'
import { shallow } from 'enzyme'
import Blog from './Blog'
import { debug } from 'util'
import { Button, Card, Icon } from 'semantic-ui-react'

describe('<Blog />', () => {

  let blogComponent
  const mockHandleDestroy = jest.fn()
  const mockHandleLike = jest.fn()

  const blog = {
    title: 'test blog',
    author: 'test author',
    url: 'http://testurl.com',
    likes: 99,
    user: { name: 'test user full name', username: 'testusername' },
    comments: []
  }

  const user = {
    username: 'testusername'
  }

  beforeEach(() => {
    blogComponent = shallow(
      <Blog
        blog={blog}
        user={user}
        handleDestroy={mockHandleDestroy}
        handleLike={mockHandleLike}
      />
    )
  })

  const findHostNodeWithText = (component, text, element) => {
    return component
      //.find('div')      
      //.findWhere(n => n.type() === element && n.text().includes(text))
      .findWhere(n => n.text().includes(text))
      .hostNodes()
  }

  it ('without valid blog redirect is returned', () => {
    const returnComponent = shallow(
      <Blog        
        user={user}
        handleDestroy={mockHandleDestroy}
        handleLike={mockHandleLike}
      />
    )
    
    expect(returnComponent.text()).toContain('<Redirect />')
  })

  it('at start displays blog author, title, url and user who added the blog', () => {
    const titleNode = blogComponent.find('.blogtitle')
    const authorNode = findHostNodeWithText(blogComponent, 'Author')
    const addedByNode = findHostNodeWithText(blogComponent, 'Added by')
    const urlNode = findHostNodeWithText(blogComponent, 'url')

    expect(titleNode.text()).toContain(blog.title)
    expect(authorNode.text()).toContain(blog.author)
    expect(authorNode.is('span')).toEqual(true)
    expect(addedByNode.text()).toContain(blog.user.name)
    expect(urlNode.text()).toContain(blog.url)
    expect(urlNode.is('a')).toEqual(true)
  })

  it('blog is shown added by anonymous when blog does not contain user information', () => {
    let anonymousBlog = { ...blog }
    delete anonymousBlog.user
    const anonymousBlogComponent = shallow(
      <Blog
        blog={anonymousBlog}
        user={user}
        handleDestroy={mockHandleDestroy}
        handleLike={mockHandleLike}
      />
    )

    const addedByNode = findHostNodeWithText(anonymousBlogComponent, 'Added by')
    expect(addedByNode.text()).toContain('anonymous')
  })

  it('clicking the like button calls like handler function', () => {
    const like = blogComponent.find('.blogVote')    
    like.at(0).simulate('click')
    expect(mockHandleLike.mock.calls.length).toBe(1)
  })

  it('delete button is shown when user has added the blog themselves', () => {
    const deleteNode = blogComponent.find('.blogDelete')        
    expect(deleteNode.prop('style')).toEqual({ display: '' }) 
  })

  it('clickling delete nutton calls handler function', () => {
    const deleteNode = blogComponent.find('.blogDelete')        
    deleteNode.simulate('click')
    expect(mockHandleDestroy.mock.calls.length).toBe(1)
  })

  it('delete button is shown on anonymous blogs', () => {
    let anonymousBlog = { ...blog }
    delete anonymousBlog.user
    const anonymousBlogComponent = shallow(
      <Blog
        blog={anonymousBlog}
        user={user}
        handleDestroy={mockHandleDestroy}
        handleLike={mockHandleLike}
      />
    )

    const deleteNode = anonymousBlogComponent.find('.blogDelete')        
    expect(deleteNode.prop('style')).toEqual({ display: '' }) 
  })

  it('delete button is hidden when user has not added the blog', () => {
    let anotherUserBlog = { ...blog, user: { name: 'someone else', username: 'someothername' } }
    const anotherUserBlogComponent = shallow(
      <Blog
        blog={anotherUserBlog}
        user={user}
        handleDestroy={mockHandleDestroy}
        handleLike={mockHandleLike}
      />
    )

    const deleteNode = anotherUserBlogComponent.find('.blogDelete')        
    expect(deleteNode.prop('style')).toEqual({ display: 'none' }) 
  })  

})