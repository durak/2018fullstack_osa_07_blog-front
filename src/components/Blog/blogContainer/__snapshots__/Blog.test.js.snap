// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Blog /> renders correcly 1`] = `
<div>
  <div
    className="ui fluid card"
    href={undefined}
    onClick={[Function]}
  >
    <div
      className="content"
    >
      <h1
        className="header"
      >
        <span
          className="blogtitle"
        >
          test blog
        </span>
        <span
          className="right floated"
        >
          <button
            className="ui negative button blogDelete"
            disabled={undefined}
            onClick={[Function]}
            role="button"
            style={
              Object {
                "display": "",
              }
            }
            tabIndex={undefined}
          >
            delete
          </button>
        </span>
      </h1>
      <div
        className="meta"
      >
        <span>
          Author: 
          test author
        </span>
      </div>
      <div
        className="meta"
      >
        <span>
          Added by: 
          test user full name
        </span>
      </div>
      <div
        className="description"
      >
        url: 
        <a
          href="http://testurl.com"
          rel="noopener noreferrer"
          target="_blank"
        >
          http://testurl.com
        </a>
      </div>
    </div>
    <div
      className="extra content"
    >
      <span>
        <i
          aria-hidden="true"
          className="like icon blogVote"
          onClick={[Function]}
        />
        99
         votes
      </span>
      <span
        className="right floated"
      >
        <i
          aria-hidden="true"
          className="comments icon"
        />
        0
         comments
      </span>
    </div>
  </div>
</div>
`;
